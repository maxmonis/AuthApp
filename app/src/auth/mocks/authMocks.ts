export const mockEmail = "user@example.com"
export const mockPassword = "Passw0rd"
export const mockToken = "mock-token"
export const mockRefreshedToken = "mock-refreshed-token"
