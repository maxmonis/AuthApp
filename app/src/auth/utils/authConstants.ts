export const accountRoutes = [
  "forgot-password",
  "login",
  "register",
  "update-password",
] as const
