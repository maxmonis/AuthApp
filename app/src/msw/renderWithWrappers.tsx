import {render} from "@testing-library/react"
import {QueryClient, QueryClientProvider} from "react-query"
import {BrowserRouter} from "react-router-dom"

const queryClient = new QueryClient({defaultOptions: {queries: {retry: false}}})

export function renderWithWrappers(element: React.ReactNode) {
  return render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{element}</BrowserRouter>
    </QueryClientProvider>,
  )
}
