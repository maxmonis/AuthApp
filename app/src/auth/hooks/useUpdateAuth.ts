import {useQueryClient} from "react-query"
import {setAuthToken} from "../utils/setAuthToken"

export function useUpdateAuth() {
  const queryClient = useQueryClient()
  return async (token: string | null) => {
    setAuthToken({token})
    if (token) await queryClient.invalidateQueries("auth")
    else queryClient.setQueryData("auth", null)
  }
}
