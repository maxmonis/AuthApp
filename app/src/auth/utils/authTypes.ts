import {accountRoutes} from "./authConstants"

export type AccountRoute = (typeof accountRoutes)[number]

export type TokenRes = {token: string}

export type UserRes = {email: string}
