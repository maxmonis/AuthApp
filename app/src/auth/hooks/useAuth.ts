import axios from "axios"
import {useQuery} from "react-query"
import {localToken} from "src/utils/storage"
import {UserRes} from "../utils/authTypes"

export function useAuth() {
  return useQuery({
    async queryFn() {
      try {
        if (!localToken.get()) throw Error("No token")
        const {data} = await axios.get<UserRes>("/api/user")
        return data
      } catch (error) {
        return null
      }
    },
    queryKey: "auth",
  })
}
