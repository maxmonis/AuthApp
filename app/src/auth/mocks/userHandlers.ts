import {rest} from "msw"
import {mockDelay} from "src/msw/mockDelay"
import {TokenRes, UserRes} from "../utils/authTypes"
import {
  mockEmail,
  mockPassword,
  mockRefreshedToken,
  mockToken,
} from "./authMocks"

export const userHandlers = [
  /* -------------------- Load User -------------------- */
  rest.get("api/user", (req, res, ctx) => {
    if (!req.headers.has("X-Auth-Token"))
      return res(ctx.status(401), ctx.json("No token"), ctx.delay(mockDelay()))
    if (
      ![mockToken, mockRefreshedToken].includes(
        req.headers.get("X-Auth-Token") ?? "",
      )
    )
      return res(
        ctx.status(401),
        ctx.json("Invalid token"),
        ctx.delay(mockDelay()),
      )
    return res(
      ctx.json({email: mockEmail} satisfies UserRes),
      ctx.delay(mockDelay()),
    )
  }),

  /* -------------------- Refresh Token -------------------- */
  rest.get("api/user/refresh", (req, res, ctx) => {
    if (!req.headers.has("X-Auth-Token"))
      return res(ctx.status(401), ctx.json("No token"), ctx.delay(mockDelay()))
    if (
      ![mockToken, mockRefreshedToken].includes(
        req.headers.get("X-Auth-Token") ?? "",
      )
    )
      return res(
        ctx.status(401),
        ctx.json("Invalid token"),
        ctx.delay(mockDelay()),
      )
    return res(
      ctx.json({token: mockRefreshedToken} satisfies TokenRes),
      ctx.delay(mockDelay()),
    )
  }),

  /* -------------------- Log In -------------------- */
  rest.post("api/user/login", async (req, res, ctx) => {
    const {email, password} = await req.json()
    if (email !== mockEmail)
      return res(
        ctx.status(404),
        ctx.json("User not found"),
        ctx.delay(mockDelay()),
      )
    if (password !== mockPassword)
      return res(
        ctx.status(401),
        ctx.json("Incorrect password"),
        ctx.delay(mockDelay()),
      )
    return res(
      ctx.json({token: mockToken} satisfies TokenRes),
      ctx.delay(mockDelay()),
    )
  }),

  /* -------------------- Create Account -------------------- */
  rest.post("api/user/create", async (req, res, ctx) => {
    const {email} = await req.json()
    if (email === mockEmail)
      return res(
        ctx.status(400),
        ctx.json("User already exists"),
        ctx.delay(mockDelay()),
      )
    return res(ctx.json("Link email sent"), ctx.delay(mockDelay()))
  }),
  rest.post("api/user/activate", async (req, res, ctx) => {
    const {token} = await req.json()
    if (token !== mockToken)
      return res(
        ctx.status(401),
        ctx.json("Invalid token"),
        ctx.delay(mockDelay()),
      )
    return res(
      ctx.json({token: mockToken} satisfies TokenRes),
      ctx.delay(mockDelay()),
    )
  }),

  /* -------------------- Update Password -------------------- */
  rest.post("api/user/update-password", async (req, res, ctx) => {
    const {email} = await req.json()
    if (email !== mockEmail)
      return res(
        ctx.status(404),
        ctx.json("User not found"),
        ctx.delay(mockDelay()),
      )
    return res(ctx.json("Password update email sent"), ctx.delay(mockDelay()))
  }),
  rest.patch("api/user/update-password", async (req, res, ctx) => {
    const {token} = await req.json()
    if (token !== mockToken)
      return res(
        ctx.status(401),
        ctx.json("Invalid token"),
        ctx.delay(mockDelay()),
      )
    return res(
      ctx.json({token: mockToken} satisfies TokenRes),
      ctx.delay(mockDelay()),
    )
  }),
]
