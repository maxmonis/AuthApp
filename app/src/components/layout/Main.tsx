import React from "react"
import {useNavigate} from "react-router-dom"
import {useAuth} from "src/auth/hooks/useAuth"
import {useUpdateAuth} from "src/auth/hooks/useUpdateAuth"
import {localDark} from "src/utils/storage"

export function Main({children}: React.PropsWithChildren) {
  const {data: user} = useAuth()
  const navigate = useNavigate()
  const updateAuth = useUpdateAuth()
  const storageDark = localDark.get()
  const [dark, setDark] = React.useState(
    storageDark === null && typeof matchMedia !== "undefined"
      ? matchMedia("(prefers-color-scheme: dark)").matches
      : storageDark ?? false,
  )
  return (
    <main {...(dark && {className: "dark"})}>
      <div className="h-screen w-screen overflow-x-hidden bg-slate-100 text-slate-950 transition-colors dark:bg-slate-950 dark:text-slate-50">
        <nav className="flex items-center justify-between gap-6 px-6 py-4">
          <button
            aria-label="toggle dark mode"
            className="text-2xl transition-transform [text-shadow:_2px_2px_2px_rgb(0_0_0_/_50%)] active:scale-90"
            onClick={() => {
              setDark(!dark)
              localDark.set(!dark)
            }}
          >
            {dark ? "ðŸŒœ" : "ðŸŒž"}
          </button>
          {user && (
            <div className="flex flex-wrap items-center justify-end gap-x-4">
              <p>{user.email}</p>
              <button
                className="underline"
                onClick={() =>
                  updateAuth(null).then(() =>
                    navigate("/login", {replace: true}),
                  )
                }
              >
                Logout
              </button>
            </div>
          )}
        </nav>
        {children}
      </div>
    </main>
  )
}
