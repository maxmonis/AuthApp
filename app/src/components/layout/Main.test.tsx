import {screen, waitFor} from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import {useAuth} from "src/auth/hooks/useAuth"
import {mockEmail, mockToken} from "src/auth/mocks/authMocks"
import {renderWithWrappers} from "src/msw/renderWithWrappers"
import {localDark, localToken} from "src/utils/storage"
import {Main} from "./Main"

let mockUser: ReturnType<typeof useAuth>["data"] = null
jest.mock("react-query", () => ({
  ...jest.requireActual("react-query"),
  useQuery: () => ({data: mockUser}),
}))

const mockNavigate = jest.fn()
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}))

function renderMain() {
  const {
    asFragment,
    container: {firstChild},
  } = renderWithWrappers(<Main>{null}</Main>)
  return {asFragment, firstChild}
}

describe("Main", () => {
  describe("dark mode", () => {
    test("button allows dark mode toggle", async () => {
      const {asFragment, firstChild} = renderMain()
      const darkModeToggle = screen.getByRole("button", {
        name: "toggle dark mode",
      })
      screen.getByText("ðŸŒž")
      expect(firstChild).not.toHaveClass("dark")
      expect(localDark.get()).toBeNull()
      expect(asFragment()).toMatchSnapshot()
      userEvent.click(darkModeToggle)
      await screen.findByText("ðŸŒœ")
      expect(firstChild).toHaveClass("dark")
      expect(localDark.get()).toStrictEqual(true)
      expect(asFragment()).toMatchSnapshot()
      userEvent.click(darkModeToggle)
      await screen.findByText("ðŸŒž")
      expect(firstChild).not.toHaveClass("dark")
      expect(localDark.get()).toStrictEqual(false)
    })
    test("uses dark mode preference from localStorage", () => {
      jest.spyOn(localDark, "get").mockReturnValueOnce(true)
      const {firstChild} = renderMain()
      screen.getByText("ðŸŒœ")
      expect(firstChild).toHaveClass("dark")
    })
    test("uses dark mode preference from browser if none in localStorage", () => {
      jest.spyOn(localDark, "get").mockReturnValueOnce(null)
      Object.defineProperty(window, "matchMedia", {
        value: jest.fn().mockImplementation(() => ({matches: true})),
      })
      const {firstChild} = renderMain()
      screen.getByText("ðŸŒœ")
      expect(firstChild).toHaveClass("dark")
    })
  })
  describe("user menu", () => {
    test("does not display logout button if no logged in user", async () => {
      renderMain()
      expect(screen.queryByRole("button", {name: "Logout"})).toBeNull()
    })
    test("displays logout button and email if user logged in", async () => {
      mockUser = {email: mockEmail}
      localToken.set(mockToken)
      const {asFragment} = renderMain()
      screen.getByText(mockEmail)
      expect(asFragment()).toMatchSnapshot()
      userEvent.click(screen.getByRole("button", {name: "Logout"}))
      await waitFor(() => {
        expect(mockNavigate).toHaveBeenCalledWith("/login", {replace: true})
      })
      expect(localToken.get()).toBeNull()
    })
  })
})
