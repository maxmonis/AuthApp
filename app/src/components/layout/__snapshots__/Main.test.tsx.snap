// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main dark mode button allows dark mode toggle 1`] = `
<DocumentFragment>
  <main>
    <div
      class="h-screen w-screen overflow-x-hidden bg-slate-100 text-slate-950 transition-colors dark:bg-slate-950 dark:text-slate-50"
    >
      <nav
        class="flex items-center justify-between gap-6 px-6 py-4"
      >
        <button
          aria-label="toggle dark mode"
          class="text-2xl transition-transform [text-shadow:_2px_2px_2px_rgb(0_0_0_/_50%)] active:scale-90"
        >
          ðŸŒž
        </button>
      </nav>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Main dark mode button allows dark mode toggle 2`] = `
<DocumentFragment>
  <main
    class="dark"
  >
    <div
      class="h-screen w-screen overflow-x-hidden bg-slate-100 text-slate-950 transition-colors dark:bg-slate-950 dark:text-slate-50"
    >
      <nav
        class="flex items-center justify-between gap-6 px-6 py-4"
      >
        <button
          aria-label="toggle dark mode"
          class="text-2xl transition-transform [text-shadow:_2px_2px_2px_rgb(0_0_0_/_50%)] active:scale-90"
        >
          ðŸŒœ
        </button>
      </nav>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Main user menu displays logout button and email if user logged in 1`] = `
<DocumentFragment>
  <main>
    <div
      class="h-screen w-screen overflow-x-hidden bg-slate-100 text-slate-950 transition-colors dark:bg-slate-950 dark:text-slate-50"
    >
      <nav
        class="flex items-center justify-between gap-6 px-6 py-4"
      >
        <button
          aria-label="toggle dark mode"
          class="text-2xl transition-transform [text-shadow:_2px_2px_2px_rgb(0_0_0_/_50%)] active:scale-90"
        >
          ðŸŒž
        </button>
        <div
          class="flex flex-wrap items-center justify-end gap-x-4"
        >
          <p>
            user@example.com
          </p>
          <button
            class="underline"
          >
            Logout
          </button>
        </div>
      </nav>
    </div>
  </main>
</DocumentFragment>
`;
